Packet++/src/VrrpLayer.cpp

```diff
@@ -0,0 +1,533 @@
+		if (packetLen <= offset) {
+			return false;
+		}
+
+		if (!shortenLayer((int)offset, packetLen - offset)) {
+			PCPP_LOG_ERROR("Cannot remove all virtual IP address(es), cannot shorted layer");
+			return false;
+		}
+
+		getPacketPtr()->ipAddrCount = 0;
+
+		return true;
+	}
+
+	void VrrpLayer::copyIPAddressToData(uint8_t *data, const IPAddress &ipAddress) const {
+		size_t ipAddressLen = getIPAddressLen();
+		if (ipAddressLen == 0) {
+			PCPP_LOG_ERROR("Cannot copy virtual IP address to data, for ip address length is invalid.");
+			return;
+		}
+
+		if (ipAddress.isIPv4()) {
+			memcpy(data, ipAddress.getIPv4().toBytes(), ipAddressLen);
+		} else if (ipAddress.isIPv6()) {
+			memcpy(data, ipAddress.getIPv6().toBytes(), ipAddressLen);
+		} else {
+		}
+	}
+
+	bool VrrpLayer::getIPAddressFromData(uint8_t *data, IPAddress &ipAddress) const {
+		if (getAddressType() == IPAddress::IPv4AddressType) {
+			IPv4Address ipv4Address(*((uint32_t *) data));
+			ipAddress = ipv4Address;
+		} else if (getAddressType() == IPAddress::IPv6AddressType) {
+			IPv6Address ipv6Address(data);
+			ipAddress = ipv6Address;
+		} else {
+			return false;
+		}
+
+		return true;
+	}
+
+	bool VrrpLayer::isIPAddressValid(IPAddress &ipAddress) const {
+		if (ipAddress.isIPv6() && (getProtocol() != VRRPv3)) {
+			PCPP_LOG_ERROR("Only VRRPv3 support IPv6 virtual address.");
+			return false;
+		}
+		if (ipAddress.getType() != getAddressType()) {
+			PCPP_LOG_ERROR("IP address version is not equal to layer's.");
+			return false;
+		}
+		if (!ipAddress.isValid()) {
+			PCPP_LOG_ERROR("IP address is invalid.");
+			return false;
+		}
+
+		return true;
+	}
+
+	IPAddress::AddressType VrrpLayer::getAddressType() const {
+		return m_addressType;
+	}
+
+	void VrrpLayer::setAddressType(IPAddress::AddressType addressType) {
+		m_addressType = addressType;
+	}
+
+	/*************
+	 * Vrrpv2Layer
+	 *************/
+
+	std::string VrrpV2Layer::getAuthTypeDesc() const {
+		std::string toStr;
+
+		toStr += "\n\tAuth Type: " + getAuthTypeDescByType(getAuthType());
+		toStr += " (";
+		toStr += std::to_string(getAuthType());
+		toStr += ")";
+
+		return toStr;
+	}
+
+	uint8_t VrrpV2Layer::getAdvInt() const {
+		auto *packet = (vrrp_packet *) getData();
+		uint16_t authAdvInt = packet->authTypeAdvInt;
+		auto *authAdvIntPtr = (vrrpv2_auth_adv *) (&authAdvInt);
+		return authAdvIntPtr->advInt;
+	}
+
+	uint8_t VrrpV2Layer::getAuthType() const {
+		auto *packet = (vrrp_packet *) getData();
+		uint16_t authAdvInt = packet->authTypeAdvInt;
+		auto *authAdvIntPtr = (vrrpv2_auth_adv *) (&authAdvInt);
+		return authAdvIntPtr->authType;
+	}
+
+	uint16_t VrrpV2Layer::fillChecksum() {
+		auto *vrrpPacket = (vrrp_packet *) getData();
+		ScalarBuffer<uint16_t> buffer = {};
+		buffer.buffer = (uint16_t *) getData();
+		buffer.len = getHeaderLen();
+
+		vrrpPacket->checksum = 0;
+		uint16_t checksum = computeChecksum(&buffer, 1);
+		vrrpPacket->checksum = htobe16(checksum);
```

Comment:
1. I'd advise renaming this method to `calculateAndSetChecksum()`
2. Instead of duplicating the code between this method and `calculateChecksum()` we can just call it:
   ```cpp
   auto *vrrpPacket = (vrrp_packet *) getData();
   vrrpPacket->checksum = htobe16(calculateChecksum());
   ```

Reply:
OK. I will fix it. After that please review again.